File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    void dfs(vector<int> & nums,vector<int> & arr,int i, int tar,vector<vector<int>> & ans){
        int n=nums.size();
        //base
        if(tar == 0){
            ans.push_back(arr);
            return;
        }
        if(tar < 0 || i == n) return;
        //logic
        arr.push_back(nums[i]);
        dfs(nums,arr,i,tar-nums[i],ans);
        arr.pop_back();
        dfs(nums,arr,i+1,tar,ans);
    }
    vector<vector<int>> combinationSum(vector<int>& nums, int tar) {
        vector<int> arr;
        vector<vector<int>> ans;
        dfs(nums,arr,0,tar,ans);
        return ans;
    }
};

File: 39-combination-sum/combination-sum.py
class Solution:
    def solve(self,i,arr,nums,tar,ans):
        if(tar == 0):
            ans.append(arr[:])
            return
        if tar <= 0 or i == len(nums):
            return
        arr.append(nums[i])
        self.solve(i,arr,nums,tar-nums[i],ans)
        arr.pop()
        self.solve(i+1,arr,nums,tar,ans)

    def combinationSum(self, nums: List[int], tar: int) -> List[List[int]]:
        arr = []
        ans = []
        self.solve(0,arr,nums,tar,ans)
        return ans